pandas/tseries/frequencies.py

```diff
@@ -295,8 +295,19 @@ def __init__(self, index, warn=True):
         if len(index) < 3:
             raise ValueError('Need at least 3 dates to infer frequency')
 
-        self.is_monotonic = (self.index.is_monotonic_increasing or
-                             self.index.is_monotonic_decreasing)
+        if not hasattr(index, "is_monotonic_increasing"):
```

Comment:
why don't you define a method on both TimedeltaArray & TimedeltaIndex which basically does the correct thing here?  I find this not very obvious what you are doing and the differencde betwee TDI and TA is pretty stark here (of the calling conventions and such).

Reply:
That's what I did originally, but @jorisvandenbossche insisted on not-that.

Reply:
>  Its a private method, not sure the reason. 

Not sure what you mean with this comment. This is a private method yes, but why would that be a reason that this is not the good place to put it? 
For now, it is also the only place where this logic is needed, so I don't think it is necessarily a bad place. Yes, if this would start to be used in a lot of places, putting it as a method on the Array class makes sense, for now I don't really care. 

What I asked was to just leave out the addition of *public* attributes on the Array classes out of this PR (as that needs a more general discussion). Whether the logic is put here, in a private attribute on the Array, or by doing a `self.index = pd.Index(self.index, copy=False)` to ensure it has the needed properties (and the rest of the code here can stay the same), I don't care much.

Reply:
because you are putting logic that clearly has a different implementation outside of the class which defines it ; this is breaking all kinds of abstraction here

the inference engine needs to know some info which is better off donenall together for perf reasons - so call a method - that is what they are for 

@jbrockmendel move this to a private method on ta and tdi

Reply:
done.  Now frequencies.infer_freq is accessing private attrs, but we can revisit that smell later